### ISSUE TITLE
Add support for tags in server browser

### ISSUE THREAD

**Issue by GeckoEidechse** [R2Northstar/Northstar#131]
(Requires support in both client/server mods and master server which is why this repository here was chosen for the issue)

# Problem

An issue I'm starting to notice is how much of the server name and server description are taken up by specifications of the server

![server list](https://user-images.githubusercontent.com/40122905/148687369-ac07d764-d9e6-4a54-bb26-6f4bf783f5db.png)

# Solution

The proposed solution would be to use **_tags_**

## Example

Basically each server can specify a list of tags like `9k air accel`, `kraber only`, `125HP` `60-tick`, etc.
Depending on the future implementation of a region selection or ping counter, these tags could also include the server region, so like `EU`, `NA West`, `AUS`, ...

This would then be displayed in the server browser in game, either below the description when clicking on the server or directly in the table on a separate column.

## Searching for tags

The main idea is to have the search bar also include searching for tags, similarly to [NorthstarMods#54](https://github.com/R2Northstar/NorthstarMods/pull/54). This way a pilot looking for a kraber 9k EU server can just put `kraber 9k EU` in the search bar and would receive all servers that include these tags.
Or someone looking for only `60-tick` servers could just put `60-tick` in the search bar and would only be shown servers with the specified tag.

Further the order of tags and the order of the search term should not be relevant, so searching for `kraber 9k EU` and `EU kraber 9k` should return the same list of servers.

## Summary

Adding support for some kind of tags that specify different attributes of the server would allow pilots to find specific servers quicker while freeing up space in server list and description.

## Discussion

A few open questions would be:
- Have a set of predefined tags (standardised list) or allow owners to specify any kind (will result in different tags with same meaning `kraber only` vs `only kraber`)
  - Middle ground: Have a standard list of tags on the wiki?
- How to deal with servers that put irrelevant tags
  - Tbh we have that issue already where servers can just set an irrelevant name/description and so far it hasn't been an issue
- Have some tags be set automatically? Like adding the current gamemode, map, and match settings as tags automatically
  - Would add more code complexity but could allow for more fine grained searches.
  - On the other side a lot of custom match settings would produce a long list of tags.
- Display some global list of tags somewhere that is the set of all tags currently used by servers. Would allow pilots to discover new niche categories but again feels like adding to much code complexity. Maybe as a standalone mod?

**Comment by MiloAkerman** [comment]
I agree completely. Would help free up clutter and account for region at least until geoloc comes around. I also like the tag list idea. I don't know about gamemode / map etc since I think those are already accounted for with the regular searchbar

**Comment by GeckoEidechse** [comment]
Regarding the potential problem of server hosters adding unrelated tags to make their servers more prominent and attract more players an issue solution would be to sort servers by least number of tags specified when filtering for certain tags.

This way server hosters are encouraged to only use the minimum number of tags to categorise their servers.

**Comment by MiloAkerman** [comment]
This would also discourage server diversity, though.

**Comment by MiloAkerman** [comment]
Now implemented with R2Northstar/NorthstarMasterServer/pull/27!

**Comment by GeckoEidechse** [comment]
Forgot to document it but when discussing this in vc the other day I think settled on tags being a list of JSON objects to allow for easier extendability in the future. Example:

```json
[
    {
        "lastHeartbeat": 123,
        "id": "abc",
        "name": "My awesome server",
        "description": "Come play!",
        "playerCount": 0,
        "maxPlayers": 12,
        "map": "mp_drydock",
        "playlist": "aitdm",
        "hasPassword": false,
        "modInfo": {
            "Mods": [
                {
                    "Name": "Northstar.Custom",
                    "Version": "1.9.0",
                    "RequiredOnClient": true,
                    "Pdiff": "// this is just an empty pdiff file so that if people roll back an update the pdiff file will be overwritten\n"
                }
            ]
        },
        "tags" :[
            {
                "category": "server-region",
                "value": "EU"
            },
            {
                "category": "TODO",
                "value": "TODO"
            },
            {
                ...
            }
        ]
    },
    ...
]
```
